Manual verification for snap yaml schema (#99)
cloud-init/ubuntu-sru: bugs/gh-364.txt

=== Begin SRU Template ===
[Impact]
When a user is configuring a snapd service, we must ensure the provided yaml conforms to
our expected schema. Furthermore, we should also verify if the new properties added to the
schema are not breaking past snap configs.

[Test Case]

```
#!/bin/sh
set -x
# Manually deploy on a lxc container

# Check if necessary scripts are found
if ! which lxc-proposed-snapshot >/dev/null; then
	echo "Cannot find lxc-proposd-snapshot, get from qa-scripts repo";
	return 1
fi

cat > snap.yaml << EOF
#cloud-config
snap:
  assertions:
    00: |
      signed_assertion_blob_here
    02: |
      signed_assertion_blob_here
  commands:
    00: snap create-user --sudoer --known <snap-user>@mydomain.com
    01: ['snap', 'install', 'canonical-livepatch']
    02: canonical-livepatch enable <AUTH_TOKEN>
EOF

> for SERIES in bionic eoan focal xenial; do
>    echo '=== BEGIN ' $SERIES
>    ref=$SERIES-proposed
>    name=test-$SERIES
>    lxc delete $name --force 2> /dev/null
>    lxc-proposed-snapshot --add-archive ppa:~cloud-init-dev/proposed --publish $SERIES $ref | egrep "Creating|cloud-init"
>    lxc launch $ref $name
>    lxc file push snap.yaml $name/var/tmp/
>    lxc file push setup_dev_proposed.sh $name/var/tmp/
>    lxc exec $name -- cloud-init status --wait --long
>    lxc exec $name -- cloud-init devel schema -c /var/tmp/snap.yaml || echo "FAILURE"
> done

Reviewer: For future, since you are really only testing updated commandline tooling here, you don't really need to use lxc-proposed-snapshot at all, just your lxc file push setup_dev_proposed.sh is sufficient.

The main benefit of using lxc-proposed-shapshot is create a base image that hasn't booted yet, which already has the updated cloud-init from the -proposed pocket.

In your case here it is unnecessary (especially because you are re-installing the setup_dev_proposed.sh version from our PPA even after you've already installed the xenial-proposed version from ubuntu proper.

It did look like in this case you pushed setup_dev_proposed.sh over to the container, but didn't run it, so we'll want to actually call it too.

And you probably also want you verification test to show the failure case before the upgrade to the new -poposed cloud-init 

Unfortunately 

   lxc exec $name -- cloud-init devel schema -c /var/tmp/snap.yaml won't actually exit 1 for you if the config is valid yaml but uncovered by any defined schema.  All it validates in that case is that the file begins with a #cloud-config header and contains valid YAML.

So in your schema tests, you'll probably want to provide an invalid-snap.yaml with invalid schema types for both assertions and commands (like providing a -<integer> value or something). assert in the first test that invalid_snap.yaml passes on current cloud-init because it is valid YAML, but happens to be invalid schema types.

Then you need to run lxc exec $name -- /var/tmp/setup_dev_proposed.sh to install -proposed version of cloud-init

And re-run cloud-init schema devel -c /var/tmp/invalid_snap.yaml to ensure it fails with the proper schema failure message, then run your valid_snap.yaml and ensure it passes.

In this test case I think the following makes sense:

```suggestion
for SERIES in bionic eoan focal xenial; do
   echo '=== BEGIN ' $SERIES
   ref=$SERIES-proposed
   name=test-$SERIES
   lxc delete $name --force 2> /dev/null
   lxc launch ubuntu-daily:$SERIES $name
   lxc file push valid_snap.yaml $name/var/tmp/
   lxc file push invalid_snap.yaml $name/var/tmp/
   lxc exec $name -- cloud-init status --wait --long   
   # add test on current cloud-init that both invalid and valid_snap.yaml pass (because we have no schema validation
    lxc exec $name -- cloud-init devel schema -c /var/tmp/invalid_snap.yaml || echo "FAILURE"
    lxc exec $name -- cloud-init devel schema -c /var/tmp/valid_snap.yaml || echo "FAILURE"
    
   # upgrade cloud-init to proposed
   lxc file push setup_dev_proposed.sh $name/var/tmp/
   lxc exec $name -- bash /var/tmp/setup_dev_proposed.sh
   # re test on current cloud-init that both invalid and valid_snap.yaml pass (because we have no schema validation
    echo "Expect schema failure on invalid_snap"
    lxc exec $name -- cloud-init devel schema -c /var/tmp/invalid_snap.yaml 
    echo "Expect success on valid_snap.yaml"    
    lxc exec $name -- cloud-init devel schema -c /var/tmp/valid_snap.yaml || echo "FAILURE"
    
    
   lxc exec $name -- bash /var/tmp/setup_dev_proposed.sh
   lxc exec $name -- cloud-init devel schema -c /var/tmp/snap.yaml || echo "FAILURE"
done
```

=== Begin SRU Template ===
[Impact]
When a user is configuring a snapd service, we must ensure the provided yaml conforms to
our expected schema. Furthermore, we should also verify if the new properties added to the
schema are not breaking past snap configs.

[Test Case]

```
#!/bin/sh
set -x
# Manually deploy on a lxc container

> # Check if necessary scripts are found
> if ! which lxc-proposed-snapshot >/dev/null; then
> 	echo "Cannot find lxc-proposd-snapshot, get from qa-scripts repo";
> 	return 1
> fi
> 

Reviewer: Not needed/used  for this script
```suggestion
```

      signed_assertion_blob_here
    02: |
      signed_assertion_blob_here
  commands:
    00: snap create-user --sudoer --known <snap-user>@mydomain.com
    01: ['snap', 'install', 'canonical-livepatch']
    02: canonical-livepatch enable <AUTH_TOKEN>
EOF

cat > invalid_snap.yaml << EOF
#cloud-config
snap:
  assertions:
    00: |
      signed_assertion_blob_here
    02: |
      signed_assertion_blob_here
  commands:
    00:
     c1: apt update
     c2: apt upgrade
    01: ['snap', 'install', 'canonical-livepatch']
    02: canonical-livepatch enable <AUTH_TOKEN>
EOF

cat > setup_dev_proposed.sh <<EOF
#/bin/bash
add-apt-repository ppa:cloud-init-dev/proposed -y
apt-get update -qq;
apt-get install -qqy cloud-init;
EOF

#for SERIES in bionic eoan focal xenial; do
for SERIES in xenial; do
   echo '=== BEGIN ' $SERIES
   ref=$SERIES-proposed
   name=test-$SERIES
   lxc delete $name --force 2> /dev/null
   lxc launch ubuntu-daily:$SERIES $name

   lxc file push valid_snap.yaml $name/var/tmp/
   lxc file push invalid_snap.yaml $name/var/tmp/

   lxc exec $name -- cloud-init status --wait --long
   lxc exec $name -- cloud-init devel schema -c /var/tmp/invalid_snap.yaml || echo "FAILURE"

>    lxc exec $name -- cloud-init devel schema -c /var/tmp/invalid_snap.yaml || echo "FAILURE"

Reviewer: ```suggestion
   lxc exec $name -- cloud-init devel schema -c /var/tmp/valid_snap.yaml || echo "FAILURE"
```
